src/main/java/training/endava/skeleton/Main.java

```diff
@@ -56,17 +60,131 @@ public static void main(String argv[]) {
 
         //Generics demo for lower-bound
 //        YellowFruits y = new YellowFruits();
+        ArrayList<Company> companies = (ArrayList<Company>) makeCompanies(Optional.empty());
+        companies = (ArrayList<Company>) populateCompanies(companies);
+
+        //Stream Exercises
+        //ex 1
+        companies.stream().forEach(System.out::println);
+
+        //ex2
+        //in person implementation (setId)
+
+        //ex3
+//        System.out.println(companies.stream().min((a,b)-> a.getFoundationYear().compareTo(b.getFoundationYear())).get().getName());
+
+        //ex4
+//        System.out.println(companies.stream().map(x -> x.getEmployeeList().size()).reduce(0,(a,b)->a+b));
+
+        //ex5
+//        String name = companies.stream()
+//                .map(x -> x.getEmployeeList()
+//                        .stream()
+//                        .map(y -> y.getName())
+//                        .reduce("", (a, b) -> a + b))
+//                .reduce("", (a, b) -> a + b);
+//        System.out.println(new Person(name, "email", "tel", 123456, companies.get(3)));
+
+        //ex6
+//        companies.stream().
+//                forEach(x -> x.getEmployeeList()
+//                        .stream()
+//                        .map(y -> y.getName())
+//                .forEach(System.out::println));
+        //ex7
+//        companies.stream()
+//                .forEach(x -> x.getEmployeeList()
+//                        .stream()
+//                        .map(y -> y.getPhoneNumber())
+//                        .forEach(System.out::println));
+//        System.out.println("distinct ----------------------------");
+//        companies.stream()
+//                .forEach(x -> x.getEmployeeList()
+//                        .stream()
+//                        .map(y -> y.getPhoneNumber())
+//                        .distinct()
+//                        .forEach(System.out::println));
+        //ex8
+//        Collector<Person, StringJoiner, String> collector =
+//                Collector.of(
+//                        () -> new StringJoiner("//", "Toate: ", " persoanele"),
+//                        (supp, person) -> supp.add(person.getName()),
+//                        StringJoiner::merge,
+//                        StringJoiner::toString
+//                );
+//        Optional<String> name = companies.stream()
+//                .map(x -> x.getEmployeeList()
+//                        .stream()
+//                        .collect(collector))
+//                .findFirst();
+//        System.out.println(name.get());
+
+        //ex9
+//        Optional<Integer> upperBound = Optional.of(new Integer(1000));
+//        ArrayList<Company> companies2 = (ArrayList<Company>) makeCompanies(upperBound);
```

Comment:
1. You don't need to cast List to ArrayList. ArrayList is an implementation of List.
2. Attributes should be of as abstract as you can, in case you want to change the implementation it has.